### ISSUE TITLE
RTree bulk updates

### ISSUE THREAD

**Issue by jexp** [neo4j-contrib/spatial#280]
Update & Merge
https://github.com/neo4j-contrib/spatial/pull/161

**Comment by phil20686** [comment]
Hi Micheal,

I just wanted to check that when you merged this you corrected the flaw in the build tree function that causes it to build a suboptimal tree. Its mentioned in my final comment there. I could possibly do it this weekend if you are not confident fixing it.

**Comment by jexp** [comment]
@phil20686 That would be amazing, sorry for the long delay. 

We would love to integrate your contribution into the latest code branch.

**Comment by jexp** [comment]
I'll give it a go tonight and will push to a branch.

**Comment by jexp** [comment]
I just had the idea for a next version: if it wouldn't be better to do the tree creation (for recreating or bulk insert) in memory and then just transfer the memory model to the graph in bulk ?

**Comment by jexp** [comment]
Worked well, updated the tests and running all now, will push soon.

**Comment by jexp** [comment]
@phil20686 regarding your comment 

> Incidentally, although this code works, there is a subtle flaw that makes the tree suboptimal. Somewhere in the build tree function it splits up the incoming nodes. iirc its in a function called partition which I added to the RTree file which attempts to presort the nodes, but the method for partitioning doesnt respect the sorting. This means that the R tree that is created has very high overlap. That was why this pull request was never merged into the main branch. I was going ot fix it but somehow I never got around to it.


I changed your list partitioning to use sub-list instead so the list is partitioned from left to right, so the order stays intact.
But then each sublist in the partition function is again sorted by XMin or YMin, shouldn't that be good enough.

It would be good to have a method that computes some statistics of the RTree like depth, child-distribution etc, to determine the quality and shape of the tree.

**Comment by jexp** [comment]
Here you go: https://github.com/neo4j-contrib/spatial/tree/bulk-update

**Comment by phil20686** [comment]
Hi I am having a think about how to test the efficiency + correctness of the RTree that gets built

Is there an easy way in cypher to test that every leaf node is the same distance from the root?

I was thinking something like 👍 

START root = node(id_of_root)
MATCH p = (root) -[:RTREE_CHILD*] ->(child) -[:RTREE_LEAF]->(leaf)
RETURN length(p)

And then check all the lengths are the same number, would that work?

**Comment by jexp** [comment]
Yes we can aggregate the lengths
And should see the distribution 

> MATCH p = (root) -[:RTREE_CHILD*] ->(child) -[:RTREE_LEAF]->(leaf)
> 
Where id(root) = id

**Comment by phil20686** [comment]
So I have been looking at this over the weekend, and I managed to uncover a bug where it sometimes throws an Inconsistent Record Exception. Says something like "This transaction was marked as a success but could not be committed because of an inconsistent record".

Do you have any insight into what causes this? Is it possibly what happens if you call node.delete() when the node still has a relationship attached?

**Comment by phil20686** [comment]
PS: I did write some code to check the overlap of the RTree and with your fix it creates a tree with no overlap.

**Comment by phil20686** [comment]
So I have uncovered a second flaw, which I am currently working on also. Sorry that this is so slow I have very little free time for this now that I work at an Evil Hedge Fund.

I have brought this branch into my fork and will pull request when I have fixed it and bulked out the test suite. While it always seems to work when you build the tree from scratch it seems to have some odd behaviours when it tries to bulk load onto an existing tree. 

You can of course look at the bulk-update in my branch of the fork to see changes.

**Comment by jexp** [comment]
Thanks so much !

I did some larger scale testing, adding 180M nodes to the RTree (doing 1M
at a time). It took 11hrs which is much much less than expected before.

I found some perf issues with querying the RTree, not sure if those are
related to the bulk addition,

Here is the result running the test query, it doesn't seem to be fully
balanced

MATCH (root) -[rels:RTREE_CHILD*] ->(child) where id(root)=247486923 and
size( (child)-[:RTREE_CHILD]->()) = 0 with size(rels) as l return
l,count(*);
+--------------+
| l | count(*) |
+--------------+
| 5 | 45549    |
| 4 | 3369701  |
| 3 | 29004    |
| 6 | 68       |
+--------------+
4 rows
12104 ms

**Comment by phil20686** [comment]
Yes that is one of the isses that I uncovered. Also, if the tree becomes unbalanced there are a bunch of things that can go wrong. I am surprised that you were able to add so many into the tree without running into an InconsistentRecord exception. I believe that that can occur when you get an unbalanced tree, as some parts of the algorithm rely on knowing how deep the tree is so that you can know where to insert.

So as for querying the tree, once the tree is build it should perform identically to any other RTree in memory. Except that these R trees have lower loading factors, that makes them more flexible under addition (virtually never have a propagating split) but slower under query as the tree is marginally deeper.

Obviously that is subject to the tree not having too much overlap. In theory these RTrees should have very little overlap...but possibly the bug leading the tree to become unbalanced means that there is significant overlap.

I wrote a utilitly that checks the balance that allows us to check the balance of any RTree before committing the transaction. Its in my fork.

**Comment by phil20686** [comment]
Also, those numbers do not appear to add up to 180M, so maybe you did run into the Inconsistent record exception?

**Comment by phil20686** [comment]
Wait, that query you gave wouldnt work would it? That gets the root also to childs which are not children on the last level.

MATCH p = (root) -[:RTREE_CHILD*] ->(child) -[:RTREE_REFERENCE]->(leaf)
    WHERE id(root) = id
    RETURN length(p) as depth, count (*) as freq

(Note that you got the relationship name wrong in your first effort to do this query above)

This should return you one relationship for each leaf node. Try that one.

**Comment by jexp** [comment]
```
MATCH p = (root) -[:RTREE_CHILD*] ->(child) -[:RTREE_REFERENCE]->(leaf)
> WHERE id(root) = 247486923
> RETURN length(p) as depth, count (*) as freq;

+-------------------+
| depth | freq      |
+-------------------+
| 5     | 223084065 |
| 4     | 2009783   |
| 7     | 2414      |
| 6     | 1618163   |
+-------------------+
4 rows
288173 ms
```

**Comment by jexp** [comment]
Which is I think, correct as there were already indexed nodes in the RTree before I added the other 170M 
But I can also re-run it with any changes you have.

**Comment by phil20686** [comment]
I am really working on trying to nail down exactly how the tree becomes unbalanced. I think that that is the root of the problem. This week is mega busy for me, but I should have a few hours over the weekend.

**Comment by jexp** [comment]
Anything I could look for or any test I could write, like adding nodes where the longitude of the new node is always close but left of the existing one?

**Comment by jexp** [comment]
I'll try to profile the lookup operation.

**Comment by phil20686** [comment]
I have a theory about what causes the imbalance, but i ahvent tested it

When adding to an existing tree, it clusters the nodes to add on the child nodes that already exist, and then asks, if i built a tree out of this cluster, how high would it be? If the answer is the "right" height, it then builds a new tree from scratch and adds these children as as children of the node on which you clustered. If the height is too small, it reclusters on the children, and repeats until it finds teh right level at which to build.

However, when inserting the children of the newly build tree, what if it causes a node to split? At that point that node can occasionally be dragged to another level by the rebalancing (v unusual as requires the tree be fully loaded right up to the root of the tree), then it would be adding to the wrong level.

An alternative possiblilty is that it simply guesses/calculates the height of the tree that it is about to build wrongly in some edges, resulting in inserting branches incorrectly.

**Comment by craigtaverner** [comment]
@phil20686 - Simon and I have been fiddling with that same part of the code, where a new subtree is created and then inserted into an existing tree, and our change gets rid of the InvalidRecordException also (an error we are seeing a lot of, and it appears from the above discussion, so are you). However, we think our chance is not necessarily correct. Please reply to me on skype and lets discuss an opportunity to compare notes on this.

**Comment by phil20686** [comment]
I think that i have fixed all of the outstanding issues. The tree was becoming unbalanced for two reasons. One the function for guessing the height of the tree did the wrong thing for a cluster of size 1, although its unclear what if anything that affected. And more importantly, the behavior of the private function partitionList function was erroneous whenever it was an exact multiple of the target loading.

int partitionSize = nodeCount / numberOfPartitions+1;

If nodeCount was eg 140 and and targetloading is 70 this would return a partition of 71 and 69, but that 71 would get an extra level now. Its critical that this function never return a partition greater than the target loading. So I changed it to:

int partitionSize = nodeCount / numberOfPartitions; 
if (nodeCount % numberOfPartitions > 0) {
       partitionSize++;
}

Which avoids this pitfall.

**Comment by phil20686** [comment]
(the plus one is because integer division can result in partionSize * numberOfPartitions < nodeCount without it)

**Comment by phil20686** [comment]
Pull request https://github.com/neo4j-contrib/spatial/pull/285

**Comment by craigtaverner** [comment]
This work was released in version 0.24 from early 2017.
